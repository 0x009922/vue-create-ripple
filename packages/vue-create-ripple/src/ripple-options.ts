import { SetRequired } from 'type-fest';

export interface RippleOptions {
  /**
   * Effect disappearence timing function
   *
   * @default 'cubic-bezier(0.4, 0, 0.2, 1)'
   */
  disappearEasing?: string;

  /**
   * Effect disappearence duration
   *
   * @default '.7s'
   */
  disappearDuration?: string;

  /**
   * Effect appearence timing function
   *
   * @default 'cubic-bezier(0.16, 1, 0.3, 1)'
   */
  appearEasing?: string;

  /**
   * Effect appearence duration
   *
   * @default '.8s'
   */
  appearDuration?: string;

  /**
   * The class to be assigned to the circle. Can be used
   * for quick and easy styling.
   *
   * ! If not specified, then nothing will be visible
   */
  class?: string | null;

  /**
   * Flag to disable effect triggering
   */
  disabled?: boolean;

  /**
   * If `true`, the effect will always come from the center of the element.
   *
   * @default false
   */
  center?: boolean;

  /**
   * If `true`, then the effect will work only on event
   * `mousedown` on the element itself, not its descendants
   *
   * @default false
   */
  self?: boolean;

  /**
   * If `true`, then the effect when clicking on the child element will not work,
   * if in the tree of its parents there is an element bound to it
   * directive and this element is a child of the current one. Default
   * `true`, so you need to explicitly specify `false` to disable this option
   *
   * @default true
   */
  noNested?: boolean;

  /**
   * Additional css styles for the container
   */
  containerStyle?: Partial<CSSStyleDeclaration>;

  /**
   * Additional css styles for the ripple
   */
  circleStyle?: Partial<CSSStyleDeclaration>;

  /**
   * The absolute value of the maximum radius of the circle (in pixels).
   * By default, the circle is stretched enough to completely cover the entire
   * element.
   *
   * `null` means to stretch
   */
  radius?: number | null;
}

export type NormalizedRippleOptions = SetRequired<
  RippleOptions,
  | 'appearDuration'
  | 'appearEasing'
  | 'center'
  | 'disabled'
  | 'disappearDuration'
  | 'disappearEasing'
  | 'noNested'
  | 'radius'
  | 'self'
>;

const defaults: NormalizedRippleOptions = {
  disappearEasing: 'cubic-bezier(0.4, 0, 0.2, 1)',
  disappearDuration: '.7s',
  appearEasing: 'cubic-bezier(0.16, 1, 0.3, 1)',
  appearDuration: '.8s',
  class: null,
  disabled: false,
  center: false,
  self: false,
  noNested: true,
  radius: null
};

export function mergeDefaults(opts?: RippleOptions): NormalizedRippleOptions {
  return { ...defaults, ...opts };
}
