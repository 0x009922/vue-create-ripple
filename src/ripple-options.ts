export interface NormalizedRippleOptions {
  /**
   * Тайминг-функция исчезновения ripple.
   * По умолчанию 'cubic-bezier(0.4, 0, 0.2, 1)'
   */
  disappearEasing: string;
  /**
   * Длительность исчезновения ripple.
   * По умолчанию '.7s'
   */
  disappearDuration: string;
  /**
   * Тайминг-функция увеличения круга.
   * По умолчанию 'cubic-bezier(0.16, 1, 0.3, 1)'
   */
  appearEasing: string;
  /**
   * Длительность увеличения круга.
   * По умолчанию '.8s'
   */
  appearDuration: string;
  /**
   * Класс, который будет задан кругу. Можно использовать
   * для быстрой и простой стилизации.
   *
   * ! Если не будет указан, то ничего не будет видно
   */
  class?: string | null;
  /**
   * Если `true`, то эффект не будет срабатывать
   */
  disabled: boolean;
  /**
   * Если `true`, то эффект всегда будет исходить из центра элемента.
   */
  center: boolean;
  /**
   * Если `true`, то эффект сработает только при событии
   * `mousedown` на самом элементе, а не его потомках
   */
  self: boolean;
  /**
   * Если `true`, то эффект при клике на дочернем элементе не сработает,
   * если в дереве его родителей найдётся элемент с привязанной к нему
   * директивой и этот элемент является дочерним к текущему. По умолчанию
   * `true`, так что нужно явно указать `false` для отключения эффекта
   */
  noNested: boolean;
  /**
   * Объект для передачи стиля контейнеру напрямую.
   */
  containerStyle?: Partial<CSSStyleDeclaration>;
  /**
   * Объект для передачи стиля круга напрямую.
   */
  circleStyle?: Partial<CSSStyleDeclaration>;
  /**
   * Абсолютное значение максимального радиуса круга (в пикселях).
   * По умолчанию круг растягивается настолько, чтобы полностью покрыть весь
   * элемент.
   */
  radius: number | null;
}

export type RippleOptions = Partial<NormalizedRippleOptions>;

const defaults: NormalizedRippleOptions = {
  disappearEasing: 'cubic-bezier(0.4, 0, 0.2, 1)',
  disappearDuration: '.7s',
  appearEasing: 'cubic-bezier(0.16, 1, 0.3, 1)',
  appearDuration: '.8s',
  class: null,
  disabled: false,
  center: false,
  self: false,
  noNested: true,
  radius: null
};

export function mergeDefaults(opts?: RippleOptions): NormalizedRippleOptions {
  return { ...defaults, ...opts };
}
