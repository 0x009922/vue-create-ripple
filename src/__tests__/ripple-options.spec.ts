import { mergeDefaults } from '../ripple-options';

describe('Ripple options', () => {
  describe('mergeDefaults()', () => {
    it('Correct merging', () => {
      expect(mergeDefaults({
        class: 'custom',
        appearDuration: '1234010s',
        circleStyle: {
          color: 'red'
        }
      })).toMatchSnapshot();
    });
  });
});
