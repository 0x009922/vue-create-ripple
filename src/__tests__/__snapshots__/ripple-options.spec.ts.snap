// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ripple options mergeDefaults() Correct merging 1`] = `
Object {
  "appearDuration": "1234010s",
  "appearEasing": "cubic-bezier(0.16, 1, 0.3, 1)",
  "center": false,
  "circleStyle": Object {
    "color": "red",
  },
  "class": "custom",
  "disabled": false,
  "disappearDuration": ".7s",
  "disappearEasing": "cubic-bezier(0.4, 0, 0.2, 1)",
  "noNested": true,
  "radius": null,
  "self": false,
}
`;
